lib/vm/src/export.rs

```diff
@@ -31,6 +31,8 @@ pub struct ExportFunction {
     pub address: *const VMFunctionBody,
     /// Pointer to the containing `VMContext`.
     pub vmctx: *mut VMContext,
+    /// temp code to set vmctx for host functions
+    pub function_ptr: Option<fn(*mut std::ffi::c_void, *const std::ffi::c_void)>,
```

Comment:
I don't understand what it does exactly, and why the name `function_ptr`? Can you explain please?

Reply:
This is `WasmerEnv::finish` for the env used by this function, due to our heirarchy of types, we can't be explicit about the type signature because this is in `vm` and it takes `&api::Instance` as an argument.

It's really not ideal but I haven't come up with a better way to get access to the initializer function yet. This is the thing I was hoping to get the most feedback on early

Reply:
I'm unclear on why we need this here. The env finisher should live in the InstanceHandle, right?

Reply:
Good question!

This is how the function pointers get into the `InstanceHandle`: we lose type information, but `wasmer::externals::function::Function` methods have access to the initializer, we have to store it somewhere for later and this is the best place I've found.

The little code window above that Ivan commented on is showing out of date code, but I left a comment there asking for feedback if anyone knows a better way to get this information to where we need it.

Reply:
Why don't store it in `InstanceHandle` ?

Reply:
We do, but we have to get that information to the `InstanceHandle`.

The data flow starts when a user makes a new `Function`:
1. [Function::new_with_env](https://github.com/wasmerio/wasmer/blob/feature/host-env-prototype/lib/api/src/externals/function.rs#L130-L161) : by way of the trait constraint, we can get the initializer and turn it into a function pointer.  We then put it in `ExportFunction` because it seemed like the most obvious place to put it.

2. We need this data when instantiating [Instance::new](https://github.com/wasmerio/wasmer/blob/feature/host-env-prototype/lib/api/src/instance.rs#L73-L76), [Module::instantiate](https://github.com/wasmerio/wasmer/blob/feature/host-env-prototype/lib/api/src/module.rs#L261-L279) , we only get a module and a resolver (the resolver is where all the functions are living)

3. We end up in [engine::Artifact::instantiate](https://github.com/wasmerio/wasmer/blob/feature/host-env-prototype/lib/engine/src/artifact.rs#L89-L158) here we extract these values from the `Imports` during instantiation and make it part of the `InstanceHandle`

4. Later we call a function on `InstanceHandle` to do something with this stored data